### DISCUSSION TITLE
Atomic CRUD

### DISCUSSION THREAD

**Discussion by SamuelMarks** [sqlalchemy/sqlalchemy#5966] [Usage Questions]
Can you critique how I'm doing CRUD with SQLalchemy?

I'm exposing the following functions on a REST API. With the exception of `D`, all should return a full entity.
```py
from sqlalchemy import Column, String, create_engine, select, update
from sqlalchemy.exc import DatabaseError
from sqlalchemy.orm import Session, declarative_base

engine = create_engine(
    # "sqlite://",
    "postgresql+psycopg2://some_user:some_pass@localhost/some_db",
    echo=True, future=True,
)

Base = declarative_base()


class Config(Base):
    __tablename__ = "config_tbl"

    dataset_name = Column(
        String, doc="name of dataset", default="mnist", primary_key=True
    )


Base.metadata.create_all(engine)


def orm_to_dict(orm_instance):
    """
    Convert instance of an SQLalchemy declarative base class into a Python dictionary

    :param orm_instance: Instance of an SQLalchemy declarative base class
    :type orm_instance: ```Base```

    :returns: Dict representation of instance
    :rtype: ```Optional[dict]```
    """
    return (
        None
        if orm_instance is None
        else {
            col.name: getattr(orm_instance, col.name)
            for col in orm_instance.__table__.columns
        }
    )
```

## `CREATE`
```py
# @rest_api.post('/api/config')
def create(config_dict):
    """
    Create `Config`

    ```yml
    requests:
        description: `Config` to add
        content:
          'application/json':
            schema:
              $ref: ```Config```

    responses:
      '201':
        description: A `Config` object.
        content:
          application/json:
            schema:
              $ref: ```Config```
      '400':
        description: A `ServerError` object.
        content:
          application/json:
            schema:
              $ref: ```ServerError```
    ```

    :param config_dict: Dictionary representation of `Config`
    :type config_dict: ```dict```

    :returns: Created ```Config``` instance (as a dict), or an error
    :rtype: ```dict```
    """
    try:
        config = Config(**config_dict)
    except TypeError as e:
        # response.status = 400
        return {"error": "ValidationError", "error_description": "\n".join(e.args)}

    try:
        with Session(engine) as session:
            session.add(config)
            session.commit()
            created_config = orm_to_dict(config)
    except DatabaseError as e:
        # response.status = 400
        return {
            "error": e.__class__.__name__,
            "error_code": e.code,
            "error_description": str(e.__cause__),
        }

    return created_config
```

## `READ`
```py
# @rest_api.get('/api/config/:dataset_name')
def read(dataset_name):
    """
    Find one `Config` or error

    ```yml
    responses:
      '200':
        description: Found `Config` object.
        content:
          application/json:
            schema:
              $ref: ```Config```
      '404':
        description: A `ServerError` object indicating no `Config` was found.
        content:
          application/json:
            schema:
              $ref: ```ServerError```
    ```

    :param dataset_name: The primary key of `Config`
    :type dataset_name: ```str```

    :returns: Found `Config` (as a dict) or error dict
    :rtype: ```dict```
    """
    with Session(engine) as session:
        config = session.execute(
            select(Config).filter_by(dataset_name=dataset_name)
        ).one_or_none()

    if config is None:
        # response.status = 404
        return {"error": "NotFound", "error_description": "Config not found"}

    return orm_to_dict(config[0])
```

## `UPDATE`
```py
# @rest_api.put('/api/config/:dataset_name')
def edit(dataset_name, config_dict):
    """
    Update one `Config` or error

    ```yml
    responses:
      '200':
        description: Updated `Config` object.
        content:
          application/json:
            schema:
              $ref: ```Config```
      '404':
        description: A `ServerError` object indicating no `Config` was found.
        content:
          application/json:
            schema:
              $ref: ```ServerError```
      '400':
        description: A `ServerError` object.
        content:
          application/json:
            schema:
              $ref: ```ServerError```
    ```

    :param dataset_name: The primary key of `Config`
    :type dataset_name: ```str```

    :param config_dict: Dictionary representation of `Config`
    :type config_dict: ```dict```

    :returns: Found `Config` (as a dict) or error dict
    :rtype: ```dict```
    """
    # config = Config(dataset_name=dataset_name)

    with Session(engine) as session:
        stmt = (
            update(Config)
            .where(Config.dataset_name == dataset_name)
            .values(config_dict)
            .returning(Config)
        )
        config = session.execute(stmt)

    if config is None:
        # response.status = 404
        return {"error": "NotFound", "error_description": "Config not found"}

    return orm_to_dict(config)
```

## `DELETE`
```py
# @rest_api.delete('/api/config/:dataset_name')
def destroy(dataset_name):
    """
    Delete one `Config`

    ```yml
    responses:
      '204':
    ```

    :param dataset_name: The primary key of `Config`
    :type dataset_name: ```str```

    :returns: None. Always. Even if nothing was deleted.
    :rtype: ```dict```
    """
    with Session(engine) as session:
        session.query(Config).filter(Config.dataset_name == dataset_name).delete(
            synchronize_session="evaluate"
        )
    # response.status = 204
```

## Test
```py
create({"dataset_name": "foo"})
read("foo")
edit("foo", {"dataset_name": "bar"})
destroy("bar")
```

**Comment by zzzeek** [comment]
It looks fine to me, though you might want to test it  - the UPDATE method is going to return a result set and not an ORM instance so you would want to accommodate that.      I just added the ability to use update and returning at the same time with full ORM objects in 1.4.0b3 if you want to try that: https://docs.sqlalchemy.org/en/14/orm/session_basics.html#orm-dml-returning-objects 

Otherwise,   while it's simplistic, I think that's absolutely appropriate if it does what you need right now.   the part with orm_to_dict() you might find would need more flexibility but I wouldn't go there until you see a need.

any comments @jvanasco ?

**Reply by jvanasco** [reply]
LGTM.

In terms of `orm_to_dict`, I usually make this a convenience method in the ORM's base class, and iterate over the `persist_selectable` columns.  That's a small detail though.

A bit of a warning though, this function will cause SQLAlchemy to load unloaded attributes.  If you want to avoid that, a paired function would be...

```python

def orm_loaded_to_dict(orm_instance):
    """
    Convert instance of an SQLalchemy declarative base class into a Python dictionary.
    Only LOADED attributes are converted. See also `orm_to_dict` which will load all attributes.

    :param orm_instance: Instance of an SQLalchemy declarative base class
    :type orm_instance: ```Base```

    :returns: Dict representation of instance
    :rtype: ```Optional[dict]```
    """
    _dict = orm_instance.__dict__
    return (
        None
        if orm_instance is None
        else {
            col.name: getattr(orm_instance, col.name)
            for col in orm_instance.__table__.columns
            if col.name in _dict
        }
    )
    ```

**Reply by SamuelMarks** [reply]
Thanks yes good point I'll add in that modification to `orm_loaded_to_dict`.

I'm still a little worried about the UPDATE, it doesn't work in SQLite, and I'm trying to create a solution that will work in any environment. Obviously the main use-case for SQLite is for testing (in mem), but there are a multitude of other use-cases that become possible when an embedded database—or a pretty-much embedded database—is used.
Also DELETE doesn't seem to actually remove the row. 

I've also played around a little bit with higher level functions, that handle all these errors to clean up the interface, e.g., `create_helper1` which returns (status_code, response): https://github.com/offscale/cdd-python/blob/master/cdd/routes/emit_constants.py

PS: I mentioned in that earlier thread which inspired you to implement the returning update that I'm writing a framework. That's kind of true. More precisely I'm writing multidirectional code-generators, that amongst other things, support both declarative_base and DDL variants of SQLalchemy: https://github.com/offscale/cdd-python

So it's kind of important that I get this CRUD right, as it'll be used in lots of places. Same for bulk CRUD.

**Reply by zzzeek** [reply]
well as you suit more use cases that will make apparent whatever limitations needs to be addressed.
